pkg/madmin/heal-commands.go

```diff
@@ -432,3 +493,141 @@ func (adm *AdminClient) HealFormat(dryrun bool) error {
 
 	return nil
 }
+
+// mkUploadsHealQuery - helper function to construct query params for
+// ListUploadsHeal API.
+func mkUploadsHealQuery(bucket, prefix, marker, uploadIDMarker, delimiter, maxUploadsStr string) url.Values {
+
+	queryVal := make(url.Values)
+	queryVal.Set("heal", "")
+	queryVal.Set(string(healBucket), bucket)
+	queryVal.Set(string(healPrefix), prefix)
+	queryVal.Set(string(healMarker), marker)
+	queryVal.Set(string(healUploadIDMarker), uploadIDMarker)
+	queryVal.Set(string(healDelimiter), delimiter)
+	queryVal.Set(string(healMaxUpload), maxUploadsStr)
+	return queryVal
+}
+
+func (adm *AdminClient) listUploadsHeal(bucket, prefix, marker, uploadIDMarker, delimiter string, maxUploads int) (listUploadsHealResponse, error) {
+	// Construct query params.
+	maxUploadsStr := fmt.Sprintf("%d", maxUploads)
+	queryVal := mkUploadsHealQuery(bucket, prefix, marker, uploadIDMarker, delimiter, maxUploadsStr)
+
+	hdrs := make(http.Header)
+	hdrs.Set(minioAdminOpHeader, "list-uploads")
+
+	reqData := requestData{
+		queryValues:   queryVal,
+		customHeaders: hdrs,
+	}
+
+	// Empty 'list' of objects to be healed.
+	toBeHealedUploads := listUploadsHealResponse{}
+
+	// Execute GET on /?heal to list objects needing heal.
+	resp, err := adm.executeMethod("GET", reqData)
+
+	defer closeResponse(resp)
+	if err != nil {
+		return listUploadsHealResponse{}, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return toBeHealedUploads, httpRespToErrorResponse(resp)
+
+	}
+
+	err = xml.NewDecoder(resp.Body).Decode(&toBeHealedUploads)
+	if err != nil {
+		return listUploadsHealResponse{}, err
+	}
+
+	return toBeHealedUploads, nil
+}
+
+// ListUploadsHeal - issues list heal uploads API request
+func (adm *AdminClient) ListUploadsHeal(bucket, prefix string, recursive bool,
+	doneCh <-chan struct{}) (<-chan UploadInfo, error) {
+
+	// Default listing is delimited at "/"
+	delimiter := "/"
+	if recursive {
+		// If recursive we do not delimit.
+		delimiter = ""
+	}
+
+	uploadIDMarker := ""
+
+	// Allocate new list objects channel.
+	uploadStatCh := make(chan UploadInfo)
```

Comment:
I think it would be good if you set the capacity of this channel to 1000, so the code will fetch the next 1000 entries before the caller of this API finishes consuming the current 1000 entries. It will be faster like that.